src/dials/command_line/refine.py

```diff
@@ -255,6 +261,71 @@ def run_macrocycle(params, reflections, experiments):
     return refiner, reflections, history
 
 
+def _find_disjoint_sets(experiments):
+
+    # Extract parameterisable models from the experiments
+    models = []
+    for experiment in experiments:
+        models.append(
+            [
+                m
+                for m in [
+                    experiment.beam,
+                    experiment.crystal,
+                    experiment.detector,
+                    experiment.goniometer,
+                ]
+                if m is not None
+            ]
+        )
+
+    # Record first set of models
+    sets = [set(models[0])]
+    ids = [
+        [0],
+    ]
+
+    # Go through all other models, matching to previous sets
+    for i, m in enumerate(models[1:]):
+        new_set = set(m)
+        disj = [new_set.isdisjoint(s) for s in sets]
+        if all(disj):
+            # no shared models, so form a new set
+            sets.append(new_set)
+            ids.append([i + 1])
```

Comment:
This maybe needs a test even if synthetic, as it is doing weird combinatorial analysis (i.e. p(🐞) >> 0)